operator/pkg/bgpv2/cluster.go

```diff
@@ -0,0 +1,229 @@
+	k8s_errors "k8s.io/apimachinery/pkg/api/errors"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	cilium_api_v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	"github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	slim_core_v1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	slim_labels "github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/labels"
+	slim_meta_v1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/meta/v1"
+)
+
+func (b *BGPResourceManager) reconcileBGPClusterConfigs(ctx context.Context) error {
+	var err error
+	for _, config := range b.clusterConfigStore.List() {
+		rcErr := b.reconcileBGPClusterConfig(ctx, config)
+		if rcErr != nil {
+			err = errors.Join(err, rcErr)
+		}
+	}
+	return err
+}
+
+func (b *BGPResourceManager) reconcileBGPClusterConfig(ctx context.Context, config *v2alpha1.CiliumBGPClusterConfig) error {
+	// get nodes which match node selector for given cluster config
+	matchingNodes, err := b.getMatchingNodes(config.Spec.NodeSelector, config.Name)
+	if err != nil {
+		return err
+	}
+
+	// update node configs for matched nodes
+	for nodeRef := range matchingNodes {
+		upsertErr := b.upsertNodeConfig(ctx, config, nodeRef)
+		if upsertErr != nil {
+			err = errors.Join(err, upsertErr)
+		}
+	}
+
+	// delete node configs for this cluster that are not in the matching nodes
+	dErr := b.deleteStaleNodeConfigs(ctx, matchingNodes, config.Name)
+	if dErr != nil {
+		err = errors.Join(err, dErr)
+	}
+
+	return err
+}
+
+func (b *BGPResourceManager) upsertNodeConfig(ctx context.Context, config *v2alpha1.CiliumBGPClusterConfig, nodeName string) error {
+	prev, exists, err := b.nodeConfigStore.GetByKey(resource.Key{Name: nodeName})
+	if err != nil {
+		return err
+	}
+
+	// find node overrides for given node
+	var overrideInstances []v2alpha1.CiliumBGPNodeConfigInstanceOverride
+	overrides := b.nodeConfigOverrideStore.List()
+	for _, override := range overrides {
+		if override.Spec.NodeRef == nodeName {
+			overrideInstances = override.Spec.BGPInstances
+			break
+		}
+	}
+
+	// create new config
+	nodeConfig := &v2alpha1.CiliumBGPNodeConfig{
+		ObjectMeta: meta_v1.ObjectMeta{
+			Name: nodeName,
+			OwnerReferences: []meta_v1.OwnerReference{
+				{
+					APIVersion: slim_core_v1.SchemeGroupVersion.String(),
+					Kind:       v2alpha1.BGPCCKindDefinition,
+					Name:       config.GetName(),
+					UID:        config.GetUID(),
+				},
+			},
+		},
+		Spec: v2alpha1.CiliumBGPNodeSpec{
+			BGPInstances: toNodeBGPInstance(config.Spec.BGPInstances, overrideInstances),
+		},
+	}
+
+	switch {
+	case exists && prev.Spec.DeepEqual(&nodeConfig.Spec):
+		return nil
+	case exists:
+		// reinitialize spec and status fields
+		prev.Spec = nodeConfig.Spec
+		_, err = b.nodeConfigClient.Update(ctx, prev, meta_v1.UpdateOptions{})
+	default:
+		_, err = b.nodeConfigClient.Create(ctx, nodeConfig, meta_v1.CreateOptions{})
+		if err != nil && k8s_errors.IsAlreadyExists(err) {
```

Comment:
good catch, I am thinking best way to tackle this would be to get latest from API server in this case and compare if spec has modified - if yes - update it otherwise it's an idempotent change.

Reply:
Makes sense,  you could probably go down the rabbit hole of worst case scenarios with this but practically those cases should be exceedingly rare.